pkg/client/william.go

```diff
@@ -0,0 +1,229 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+)
+
+// William Interface
+type William interface {
+	Enable()
+	SetClient(client HttpClient)
+	SetKey(id, secret string)
+	GetServiceName() string
+	ListPermission(ctx context.Context, ownershipLevel, action string, resourceHierarchy ...string) ([]byte, error)
+	HandlerFunc(permission func(r *http.Request) string, next http.HandlerFunc) http.HandlerFunc
+	AmHandler(w http.ResponseWriter, r *http.Request)
+	AddAction(action string)
+	AddActionFunc(action string, f ActionResourceFunc)
+	Generate(ownershipLevel, action string, resourceHierarchy ...string) func(r *http.Request) string
+}
+
+type AmPermission struct {
+	Alias    string `json:"alias,omitempty"`
+	Prefix   string `json:"prefix"`
+	Complete bool   `json:"complete"`
+}
+
+type ActionResourceFunc func(context.Context, string) []AmPermission
+
+type HttpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type william struct {
+	sync.RWMutex
+	actions map[string]ActionResourceFunc
+
+	keyID     string
+	keySecret string
+
+	serviceName string
+	baseURL     string
+	bypass      bool
+
+	client HttpClient
+}
+
+func New(baseURL, serviceName string) William {
+	wi := &william{
+		baseURL:     baseURL,
+		actions:     make(map[string]ActionResourceFunc),
+		serviceName: serviceName,
+		bypass:      true,
+		client:      http.DefaultClient,
+	}
+
+	return wi
+}
+
+func (wi *william) Enable() {
+	wi.bypass = false
+}
+
+func (wi *william) SetClient(client HttpClient) {
+	wi.Lock()
+	defer wi.Unlock()
+
+	wi.client = client
+}
+
+func (wi *william) SetKey(id, secret string) {
+	wi.keyID = id
+	wi.keySecret = secret
+}
+
+func (wi *william) GetServiceName() string { return wi.serviceName }
+
+func (wi *william) hasPermission(ctx context.Context, accessToken, permission string) (*AuthInfo, error) {
+	if len(accessToken) < 8 {
+		return nil, errors.New("invalid auth token")
+	}
+
+	auth, err := wi.get(ctx, "/permissions/has?permission="+url.QueryEscape(permission), accessToken)
+	if err != nil {
+		return nil, err
+	}
+
+	auth.permission = permission
+	return auth, err
+}
+
+func (wi *william) ListPermission(ctx context.Context, ownershipLevel, action string, resourceHierarchy ...string) ([]byte, error) {
+	params := url.Values{}
+	permission := append([]string{wi.serviceName, ownershipLevel, action}, resourceHierarchy...)
+	params.Set("permission", strings.Join(permission, "::"))
+	params.Set("pageSize", "0")
+
+	authToken := fmt.Sprintf("KeyPair %s:%s", wi.keyID, wi.keySecret)
+	authInfo, err := wi.get(ctx, "/service_accounts/with_permission?"+params.Encode(), authToken)
+	if err != nil {
+		return nil, err
+	}
+
+	return authInfo.body, nil
+}
+
+func (wi *william) get(ctx context.Context, path, authtoken string) (*AuthInfo, error) {
```

Comment:
`authtoken` is the same thing as the `access token` above? If yes, we should keep the names consistent.